File: C++_Version/Array/53Maximum_Subarray.cpp
//Find the contiguous subarray within an array(containing at least one number) which has the largest sum.
//
//For example, given the array[-2, 1, -3, 4, -1, 2, 1, -5, 4],
//the contiguous subarray[4, -1, 2, 1] has the largest sum = 6.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
using namespace std;

class Solution {
public:
	int maxSubArray(vector<int>& nums) {
		int n = nums.size();

		int sum = 0, ans = nums[0];

		for (int i = 0; i < n; ++i)
		{
			sum += nums[i];
			ans = max(sum, ans);
			sum = max(sum, 0);
		}

		return ans;
	}
};


int main()
{
	vector<int> vec1({ -2, 1, -3, 4, -1, 2, 1, -5, 4 });

	Solution sol1;
	int ret1 = sol1.maxSubArray(vec1);

	/*for (int i = 0; i < (int)number.size(); ++i)
		cout << number[i] << ' ';*/

	system("pause");

	return 0;
}

File: Java_Version/53_Maximum_Subarray.java
/*
  Find the contiguous subarray within an array (containing at least one number)
which has the largest sum.

  For example, given the array [-2,1,-3,4,-1,2,1,-5,4], the contiguous subarray
[4,-1,2,1] has the largest sum = 6.
*/

class Solution {

    public static void main(String[] args) {

        System.out.println(new Solution().maxSubArray(
                new int[]{-2, 1, -3, 4, -1, 2, 1, -5, 4}));
    }

    public int maxSubArray(int[] nums) {

        if (nums == null || nums.length == 0) {
            return 0;
        }

        int maxSum = Integer.MIN_VALUE;
        int lastMaxSum = 0;

        for (int num : nums) {

            lastMaxSum += num;
            maxSum = Math.max(maxSum, lastMaxSum);

            if (lastMaxSum < 0) {
                lastMaxSum = 0;
            }
        }

        return maxSum;
    }
}
